draft-ietf-mls-protocol.md

```diff
@@ -1000,6 +1000,70 @@ struct {
 } UserInitKey;
 ~~~~~
 
+# Message Framing
+
+[[ TODO prose ]]
+
+~~~~~
+enum {
+    invalid(0),
+    handshake(1),
+    application(2),
+    (255)
+} ContentType;
+
+struct {
+    opaque content[MLSPlaintext.length];
+    uint8 signature[MLSInnerPlaintext.sig_len];
+    uint16 sig_len;
+    ContentType type;
+    uint8 zero\_padding[length\_of\_padding];
+} MLSInnerPlaintext;
+
+struct {
+    opaque group_id<0..255>;
+    uint32 epoch;
+    uint32 sender;
+    ContentType type;
+
+    select (MLSPlaintext.type) {
+        case handshake:   Handshake;
+        case application: Application;
+    }
+
+    opaque signature<1..2^16-1>;
+} MLSPlaintext;
+
+struct {
+    opaque group_id<0..255>;
+    uint32 epoch;
+    uint32 sender;
+    uint32 generation;
```

Comment:
While this is weakly-authenticated via the AEAD it should be in the MLSPlaintext header as well and covered by the signature (especially in the case of an Application message) to avoid forgery from an adversarial member who is not the sender.

Reply:
It is also worth noting that this field even if not technically needed in the case of a Handshake message might be able to use this to advertise a counter for application messages previously sent by the sender, hence allowing recipients to know if they received all app messages from the sender. We could limit that to the previous epoch or not... Maybe there is something there or not...

Reply:
What is the generation field for exactly?